File: 백준/Silver/17484. 진우의 달 여행 （Small）/진우의 달 여행 （Small）.py
n, m = map(int, input().split())
cost = [list(map(int, input().split())) for _ in range(n)]
dp = [[[0] * 3 for _ in range(m)] for _ in range(n)]

for i in range(m):
    for j in range(3):
        dp[0][i][j] = cost[0][i]

for i in range(1, n):
    for j in range(m):
        for k in range(3):
            if (j == 0 and k == 0) or (j == m-1 and k == 2):
                dp[i][j][k] = 1000000
            elif k == 0:
                dp[i][j][0] = cost[i][j] + min(dp[i-1][j-1][1], dp[i-1][j-1][2])
            elif k == 1:
                dp[i][j][1] = cost[i][j] + min(dp[i-1][j][0], dp[i-1][j][2])
            elif k == 2:
                dp[i][j][2] = cost[i][j] + min(dp[i-1][j+1][0], dp[i-1][j+1][1])

result = 1000000
for i in dp[-1]:
    for j in i:
        result = min(result, j)
print(result)

File: 백준/Silver/17484. 진우의 달 여행 （Small）/진우의 달 여행 （Small）.c
#include <stdio.h>
#include <stdlib.h>

int min(int a, int b) {
    if (a > b) { return b; }
    else { return a; }
}

int main(void) {
    int n, m;       // 행, 열
    scanf("%d %d", &n, &m);

    int **arr = (int **)malloc(n * sizeof(int *));      // 행
    for (int i=0; i < n; i++) {         // 열
        arr[i] = (int *)malloc(m * sizeof(int));
    }

    for (int i=0; i < n; i++) {
        for (int j=0; j < m; j++) {
            scanf("%d", &arr[i][j]);
        }
    }

    int ***dp = (int ***)malloc(n * sizeof(int **));      // 행
    for (int i=0; i < n; i++) {         // 열
        dp[i] = (int **)malloc(m * sizeof(int *));

        for (int j=0; j < m; j++) {
            dp[i][j] = (int *)malloc(3 * sizeof(int));
        }
    }

    for (int i=0; i<m; i++) {
        for (int j=0; j < 3; j++) {
            dp[0][i][j] = arr[0][i];
        }
    }

    for (int i=1; i < n; i++) {
        for (int j=0; j < m; j++) {
            for (int k=0; k < 3; k++) {
                if ((j == 0 && k == 0) || (j == m-1 && k == 2)) {
                    dp[i][j][k] = 1e06;
                }
                else if (k == 0) {
                    dp[i][j][0] = arr[i][j] + min(dp[i-1][j-1][1], dp[i-1][j-1][2]);
                }
                else if (k == 1) {
                    dp[i][j][1] = arr[i][j] + min(dp[i-1][j][0], dp[i-1][j][2]);
                }
                else if (k == 2) {
                    dp[i][j][2] = arr[i][j] + min(dp[i-1][j+1][0], dp[i-1][j+1][1]);
                }
            }
        }
    }

    int result = 1e06;
    for (int i=0; i < m; i++) {
        for (int j=0; j < 3; j++) {
            result = min(result, dp[n-1][i][j]);
        }
    }

    printf("%d", result);

    return 0;
}